clientconn.go

```diff
@@ -196,9 +196,9 @@ func WithTimeout(d time.Duration) DialOption {
 }
 
 // WithDialer returns a DialOption that specifies a function to use for dialing network addresses.
-func WithDialer(f func(addr string, timeout time.Duration) (net.Conn, error)) DialOption {
+func WithDialer(d net.Dialer) DialOption {
```

Comment:
WithDialer(net.Dialer) is not sufficient to complete the goal in some cases, e.g.,
1. https://github.com/grpc/grpc-go/issues/639;
2. ppl use https://github.com/golang/appengine/blob/master/socket/socket_vm.go#L32 to dial to appengine;
3. We may introduce in-process server later which cannot be dialed with a net.Dialer either.